src/server/pps/server/githook/githook.go

```diff
@@ -0,0 +1,184 @@
+}
+func matchingBranch(inputBranch string, payloadBranch string) bool {
+	if inputBranch == payloadBranch {
+		return true
+	}
+	if inputBranch == "" && payloadBranch == "master" {
+		return true
+	}
+	return false
+}
+
+func (s *gitHookServer) findMatchingPipelineInputs(payload github.PushPayload) (pipelines []*pps.PipelineInfo, inputs []*pps.GitInput, err error) {
+	var walkInput func(*pps.Input)
+	payloadBranch := getBranch(payload.Ref)
+	pipelines, err = s.client.ListPipeline()
+	if err != nil {
+		return nil, nil, err
+	}
+	for _, pipelineInfo := range pipelines {
+		pps.VisitInput(pipelineInfo.Input, func(input *pps.Input) {
+			if input.Git != nil {
+				if input.Git.URL == payload.Repository.CloneURL && matchingBranch(input.Git.Branch, payloadBranch) {
+					inputs = append(inputs, input.Git)
+				}
+			}
+			var inputs []*pps.Input
+			if input.Cross != nil {
+				inputs = input.Cross
+			}
+			if input.Union != nil {
+				inputs = input.Union
+			}
+			for _, input := range inputs {
+				walkInput(input)
+			}
+		})
+	}
+	if len(inputs) == 0 {
+		return nil, nil, fmt.Errorf("no pipeline inputs corresponding to github URL (%v) on branch (%v) found, perhaps the github input is not set yet on a pipeline", payload.Repository.CloneURL, payloadBranch)
+	}
+	return pipelines, inputs, nil
+}
+
+func (s *gitHookServer) HandlePush(payload interface{}, header webhooks.Header) {
+	var err error
+	defer func() {
+		// handle any return error
+		if err != nil {
+			logrus.Errorf("github webhook failed to handle push with error: %v\n", err)
+		}
+	}()
+
+	pl := payload.(github.PushPayload)
+	logrus.Infof("received github push payload for repo (%v) on branch (%v)", pl.Repository.Name, getBranch(pl.Ref))
+
+	raw, err := json.Marshal(pl)
+	if err != nil {
+		err = fmt.Errorf("error marshalling payload (%v): %v", pl, err)
+		return
+	}
+	pipelines, gitInputs, err := s.findMatchingPipelineInputs(pl)
+	if err != nil {
+		return
+	}
+	if pl.Repository.Private {
+		var loopErr error
+		for _, pipelineInfo := range pipelines {
+			someErr := util.FailPipeline(context.Background(), s.etcdClient, s.pipelines, pipelineInfo.Pipeline.Name, fmt.Sprintf("unable to clone private github repo (%v)", pl.Repository.CloneURL))
+			// Err will be handled by final defer function, but first we want to
+			// try and fail all relevant pipelines
+			if someErr != nil {
+				// Only set loopErr if someErr isn't nil
+				loopErr = someErr
+			}
+		}
+		err = loopErr
+		return
+	}
+	triggeredRepos := make(map[string]bool)
+	for _, input := range gitInputs {
+		repoName := pps.RepoNameFromGitInfo(input.URL, input.Name)
+		if alreadyTriggered := triggeredRepos[repoName]; alreadyTriggered {
+			// This input is used on multiple pipelines, and we've already
+			// committed to this input repo
+			continue
+		}
+		branchName := "master"
+		if input.Branch != "" {
+			branchName = input.Branch
+		}
```

Comment:
You shouldn't be implementing default values here, implement defaults in one place.